File: MHAudioBufferPlayer/MHAudioBufferPlayer.m

#import "MHAudioBufferPlayer.h"

// The number of Audio Queue buffers we keep in rotation
#define NumberOfAudioDataBuffers 3

@interface MHAudioBufferPlayer ()

@property (nonatomic, assign, readwrite) BOOL playing;

- (void)setUpAudio;
- (void)tearDownAudio;
- (void)start;

@end

static void InterruptionListenerCallback(void *inUserData, UInt32 interruptionState)
{
	MHAudioBufferPlayer *player = (__bridge MHAudioBufferPlayer *)inUserData;
	if (interruptionState == kAudioSessionBeginInterruption)
	{
		[player tearDownAudio];
	}
	else if (interruptionState == kAudioSessionEndInterruption)
	{
		[player setUpAudio];
		[player start];
	}
}

static void PlayCallback(void *inUserData, AudioQueueRef inAudioQueue, AudioQueueBufferRef inBuffer)
{
	MHAudioBufferPlayer *player = (__bridge MHAudioBufferPlayer *)inUserData;
	if (player.playing && player.block != nil)
	{
		player.block(inBuffer, player.audioFormat);
		AudioQueueEnqueueBuffer(inAudioQueue, inBuffer, 0, NULL);
	}
}

@implementation MHAudioBufferPlayer
{
	// the audio queue object being used for playback
	AudioQueueRef _playQueue;
	
	// the audio queue buffers for the playback audio queue
	AudioQueueBufferRef _playQueueBuffers[NumberOfAudioDataBuffers];

	// the number of audio data packets to use in each audio queue buffer
	UInt32 _packetsPerBuffer;

	// the number of bytes to use in each audio queue buffer
	UInt32 _bytesPerBuffer;
}

- (id)initWithSampleRate:(Float64)sampleRate channels:(UInt32)channels bitsPerChannel:(UInt32)bitsPerChannel secondsPerBuffer:(Float64)secondsPerBuffer
{
	return [self initWithSampleRate:sampleRate channels:channels bitsPerChannel:bitsPerChannel packetsPerBuffer:(UInt32)(secondsPerBuffer * sampleRate)];
}

- (id)initWithSampleRate:(Float64)sampleRate channels:(UInt32)channels bitsPerChannel:(UInt32)bitsPerChannel packetsPerBuffer:(UInt32)packetsPerBuffer
{
	if ((self = [super init]))
	{
		_playing = NO;
		_playQueue = NULL;
		_gain = 1.0;

		_audioFormat.mFormatID         = kAudioFormatLinearPCM;
		_audioFormat.mSampleRate       = sampleRate;
		_audioFormat.mChannelsPerFrame = channels;
		_audioFormat.mBitsPerChannel   = bitsPerChannel;
		_audioFormat.mFramesPerPacket  = 1;  // uncompressed audio
		_audioFormat.mBytesPerFrame    = _audioFormat.mChannelsPerFrame * _audioFormat.mBitsPerChannel/8;
		_audioFormat.mBytesPerPacket   = _audioFormat.mBytesPerFrame * _audioFormat.mFramesPerPacket;
		_audioFormat.mFormatFlags      = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked;

		_packetsPerBuffer = packetsPerBuffer;
		_bytesPerBuffer = _packetsPerBuffer * _audioFormat.mBytesPerPacket;

		[self setUpAudio];
	}
	return self;
}

- (void)dealloc
{
	[self tearDownAudio];
}

- (void)setUpAudio
{
	if (_playQueue == NULL)
	{
		[self setUpAudioSession];
		[self setUpPlayQueue];
		[self setUpPlayQueueBuffers];
	}
}

- (void)tearDownAudio
{
	if (_playQueue != NULL)
	{
		[self stop];
		[self tearDownPlayQueue];
		[self tearDownAudioSession];
	}
}

- (void)setUpAudioSession
{
	AudioSessionInitialize(
		NULL,
		NULL,
		InterruptionListenerCallback,
		(__bridge void *)self);

	UInt32 sessionCategory = kAudioSessionCategory_MediaPlayback;
	AudioSessionSetProperty(
		kAudioSessionProperty_AudioCategory,
		sizeof(sessionCategory),
		&sessionCategory);

	AudioSessionSetActive(true);
}

- (void)tearDownAudioSession
{
	AudioSessionSetActive(false);
}

- (void)setUpPlayQueue
{
	AudioQueueNewOutput(
		&_audioFormat,
		PlayCallback,
		(__bridge void *)self, 
		NULL,                   // run loop
		kCFRunLoopCommonModes,  // run loop mode
		0,                      // flags
		&_playQueue);

	self.gain = 1.0;
}

- (void)tearDownPlayQueue
{
	AudioQueueDispose(_playQueue, YES);
	_playQueue = NULL;
}

- (void)setUpPlayQueueBuffers
{
	for (int t = 0; t < NumberOfAudioDataBuffers; ++t)
	{
		AudioQueueAllocateBuffer(
			_playQueue,
			_bytesPerBuffer,
			&_playQueueBuffers[t]);
	}
}

- (void)primePlayQueueBuffers
{
	for (int t = 0; t < NumberOfAudioDataBuffers; ++t)
	{
		PlayCallback((__bridge void *)self, _playQueue, _playQueueBuffers[t]);
	}
}

- (void)start
{
	if (!self.playing)
	{
		self.playing = YES;
		[self primePlayQueueBuffers];
		AudioQueueStart(_playQueue, NULL);
	}
}

- (void)stop
{
	if (self.playing)
	{
		AudioQueueStop(_playQueue, TRUE);
		self.playing = NO;
	}
}

- (void)setGain:(Float32)gain
{
	if (_gain != gain)
	{
		_gain = gain;
		AudioQueueSetParameter(_playQueue, kAudioQueueParam_Volume, _gain);
	}
}

@end

File: Demo-Swift/Demo-Swift/MHAudioBufferPlayer/MHAudioBufferPlayer.swift
//
//  MHAudioBufferPlayer.swift
//  Demo-Swift
//
//  Created by Kelvin Lau on 2018-01-11.
//  Copyright © 2018 Kelvin Lau. All rights reserved.
//

import AudioToolbox
import AVFoundation

typealias MHAudioBufferPlayerBlock = (_ buffer: AudioQueueBufferRef, _ audioFormat: AudioStreamBasicDescription) -> Void
let NumberOfAudioDataBuffers = 3

func InterruptionListenerCallback(inUserData: UnsafeRawPointer, interruptionState: UInt32) {
  let player = Unmanaged<MHAudioBufferPlayer>.fromOpaque(inUserData).takeUnretainedValue()
  if interruptionState == kAudioSessionBeginInterruption {
    player.tearDownAudio()
  } else if interruptionState == kAudioSessionEndInterruption {
    player.setUpAudio()
    player.start()
  }
}

func PlayCallback(_ inUserData: UnsafeMutableRawPointer!, _ inAudioQueue: AudioQueueRef, _ inBuffer: AudioQueueBufferRef) {
  let player = Unmanaged<MHAudioBufferPlayer>.fromOpaque(inUserData).takeUnretainedValue()
  if player.playing && player.block != nil {
    player.block!(inBuffer, player.audioFormat)
    AudioQueueEnqueueBuffer(inAudioQueue, inBuffer, 0, nil)
  }
}

class MHAudioBufferPlayer {
  
  var block: MHAudioBufferPlayerBlock?
  private(set) var playing: Bool
  var gain: Float32
  var audioFormat: AudioStreamBasicDescription
  
  var _playQueue: AudioQueueRef?
  var _playQueueBuffers: [AudioQueueBufferRef?] = .init(repeating: nil, count: NumberOfAudioDataBuffers)
  var _packetsPerBuffer: UInt32
  var _bytesPerBuffer: UInt32
  
  convenience init(sampleRate: Float64, channels: UInt32, bitsPerChannel: UInt32, secondsPerBuffer: Float64) {
    self.init(sampleRate: sampleRate, channels: channels, bitsPerChannel: bitsPerChannel, packetsPerBuffer: UInt32(secondsPerBuffer * sampleRate))
  }
  
  init(sampleRate: Float64, channels: UInt32, bitsPerChannel: UInt32, packetsPerBuffer: UInt32) {
    playing = false
    _playQueue = nil
    gain = 1.0
    
    audioFormat = AudioStreamBasicDescription()
    
    audioFormat.mFormatID = kAudioFormatLinearPCM
    audioFormat.mSampleRate = sampleRate
    audioFormat.mChannelsPerFrame = channels
    audioFormat.mBitsPerChannel = bitsPerChannel
    audioFormat.mFramesPerPacket = 1 // uncompressed audio
    audioFormat.mBytesPerFrame = audioFormat.mChannelsPerFrame * audioFormat.mBitsPerChannel / 8
    audioFormat.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked
    
    _packetsPerBuffer = packetsPerBuffer
    _bytesPerBuffer = _packetsPerBuffer * audioFormat.mBytesPerPacket
    
    setUpAudio()
  }
  
  deinit {
    tearDownAudio()
  }
  
  func setUpAudio() {
    if _playQueue == nil {
      setUpAudioSession()
      setUpPlayQueue()
      setUpPlayQueueBuffers()
    }
  }
  
  func tearDownAudio() {
    if _playQueue != nil {
      stop()
      tearDownPlayQueue()
      tearDownAudioSession()
    }
  }
  
  func setUpAudioSession() {
    
    // TODO: - Verify that `AudioQueueInitialize` is not necessary.
    
    do {
      try AVAudioSession.sharedInstance().setActive(true)
    } catch {
      print("Audio session setup failed: \(error)")
    }
  }
  
  func tearDownAudioSession() {
    do {
      try AVAudioSession.sharedInstance().setActive(false)
    } catch {
      print("Audio session teardown failed: \(error)")
    }
  }
  
  func setUpPlayQueue() {
    AudioQueueNewOutput(
      &audioFormat,
      PlayCallback,
      UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque()),
      nil,
      (CFRunLoopMode.commonModes as! CFString),
      0,
      &_playQueue)
    
    gain = 1.0
  }
  
  func tearDownPlayQueue() {
    AudioQueueDispose(_playQueue!, true)
    _playQueue = nil
  }
  
  func setUpPlayQueueBuffers() {
    for t in stride(from: 0, to: NumberOfAudioDataBuffers, by: 1) {
      AudioQueueAllocateBuffer(_playQueue!, _bytesPerBuffer, &_playQueueBuffers[t])
    }
  }
  
  func primePlayQueueBuffers() {
    for t in stride(from: 0, to: NumberOfAudioDataBuffers, by: 1) {
      PlayCallback(Unmanaged<MHAudioBufferPlayer>.passUnretained(self).toOpaque(), _playQueue!, _playQueueBuffers[t]!)
    }
  }
  
  func start() {
    if !playing {
      playing = true
      primePlayQueueBuffers()
      AudioQueueStart(_playQueue!, nil)
    }
  }
  
  func stop() {
    if playing {
      AudioQueueStop(_playQueue!, true)
      playing = false
    }
  }
  
  func set(gain: Float32) {
    self.gain = gain
    AudioQueueSetParameter(_playQueue!, kAudioQueueParam_Volume, gain)
  }
}


